### PR TITLE
Socket bridge

### PR DESCRIPTION
**Pull request by AlbertSu123**

### CODE REVIEW

File: src/actions/Bridge.ts [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+        fromChainId,
+        await toChain.getChainId(),
+        await fromTokenObj.getAddress(),
+        await toTokenObj.getAddress(),
+        amount,
+        sort
+    )
+    const socketTx = await getSocketBridgeTransaction(route)
+    const { allowanceTarget, minimumApprovalAmount } = socketTx.result.approvalData
+    if (socketTx.result.approvalData !== null) {
```
Comment (Chazzzzzzz) [inline src/actions/Bridge.ts:43]: if null, then return { to: socketTx.result.txTarget, value: socketTx.result.value, data: socketTx.result.txData }?

maybe use this style

if (null) {
  return XXX
} else {
  xxx
}

File: src/actions/types.ts [hunk 2]
```diff
@@ -153,3 +153,13 @@ export type OneInchSwapParams = {
     allowPartialFill: boolean // Set to true to allow partial filling of the swap order
     chainId: number // Chain ID of the blockchain to use (e.g., 1 for Ethereum Mainnet)
 }
+
+export type BridgeParams = {
+    fromChain: number
```
Comment (Chazzzzzzz) [inline src/actions/types.ts:158]: can we do string | Chain | number for fromChain and toChain?

Comment (Chazzzzzzz) [reply src/actions/types.ts:158]: also pls do the same thing for OneInchSwapParams

File: src/actions/types.ts [hunk 3]
```diff
@@ -153,3 +153,13 @@ export type OneInchSwapParams = {
...
+    toChain: number
+    fromToken: string
+    toToken: string
+    amount: number
+    sort?: "output" | "gas" | "time"
```
Comment (Chazzzzzzz) [inline src/actions/types.ts:163]: are those all possible values?
can we make an enum

Comment (Chazzzzzzz) [reply src/actions/types.ts:163]: if not set, do we have a default value for this?

Comment (Chazzzzzzz) [reply src/actions/types.ts:163]: Seems API side always require this field, so we need to set a default value

File: src/apis/BridgeApis.ts [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+    const params = new URLSearchParams({
+        recipient: recipient,
+        fromChainId: fromChain,
+        fromTokenAddress: fromToken,
+        toChainId: toChain,
+        toTokenAddress: toToken,
+        fromAmount: amount.toString(),
+        userAddress: walletAddress,
+        uniqueRoutesPerBridge: "true",
+        sort: sort ?? "output",
```
Comment (Chazzzzzzz) [inline src/apis/BridgeApis.ts:25]: lets set the default value of parameters in src/actions/Bridge.ts

API method should just take input and call apis

File: src/apis/BridgeApis.ts [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+            { walletAddress, fromChain, toChain, fromToken, toToken, amount, sort },
+            "Try another route with a different asset pair",
+            "https://docs.fun.xyz"
+        )
+    }
+    const route = quote.result.routes[0]
+    return route
+}
+
+export async function getSocketBridgeTransaction(route: any): Promise<any> {
```
Comment (Chazzzzzzz) [inline src/apis/BridgeApis.ts:42]: how hard it is to define a type for the result?

File: src/apis/BridgeApis.ts [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+    return await sendPostRequest(API_URL, "bridge/build-tx/", { route })
+}
+
+export async function getSocketBridgeAllowance(chainId: string, sender: Address, allowanceTarget: string, token: string): Promise<any> {
```
Comment (Chazzzzzzz) [inline src/apis/BridgeApis.ts:46]: how hard it is to define a type for the result?

File: src/apis/BridgeApis.ts [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+    return json
+}
+
+export async function getSocketBridgeApproveTransaction(
+    chainId: string,
+    sender: Address,
+    allowanceTarget: string,
+    token: string,
+    amount: bigint
+): Promise<any> {
```
Comment (Chazzzzzzz) [inline src/apis/BridgeApis.ts:72]: how hard it is to define a type for the result?

File: tests/testUtils/Bridge.ts [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+            const userOp = await wallet.bridge(auth, await auth.getAddress(), {
+                fromChain: config.fromChainId,
+                toChain: config.toChainId,
+                fromToken: config.fromToken,
+                toToken: config.toToken,
+                amount: config.amountToBridge,
+                sort: "gas"
+            })
+            const bridgeExecutionReceipt = await wallet.executeOperation(auth, userOp)
+            expect(bridgeExecutionReceipt).to.not.throw
```
Comment (Chazzzzzzz) [inline tests/testUtils/Bridge.ts:75]: can we check that the token is indeed received on toChain?

Comment (Chazzzzzzz) [reply tests/testUtils/Bridge.ts:75]: also couple failure cases about "not enough token", wrong token address something like that

File: src/actions/types.ts (hunk 3 above)
Comment (AlbertSu123) [reply src/actions/types.ts:163]: fixed

File: src/apis/BridgeApis.ts (hunk 7 above)
Comment (AlbertSu123) [reply src/apis/BridgeApis.ts:72]: very difficult, we would need to define the socket json format

File: src/apis/BridgeApis.ts (hunk 6 above)
Comment (AlbertSu123) [reply src/apis/BridgeApis.ts:46]: very difficult, we would need to define the socket json format

File: src/apis/BridgeApis.ts (hunk 5 above)
Comment (AlbertSu123) [reply src/apis/BridgeApis.ts:42]: very difficult, we would need to define the socket json format

File: src/apis/BridgeApis.ts (hunk 4 above)
Comment (AlbertSu123) [reply src/apis/BridgeApis.ts:25]: fixed

File: tests/testUtils/Bridge.ts (hunk 8 above)
Comment (AlbertSu123) [reply tests/testUtils/Bridge.ts:75]: we can't check, otherwise the tests would take 10+ minutes to run

Comment (AlbertSu123) [reply tests/testUtils/Bridge.ts:75]: added error tests

File: src/actions/Bridge.ts [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+        fromChainId,
+        await toChain.getChainId(),
+        await fromTokenObj.getAddress(),
+        await toTokenObj.getAddress(),
+        amount,
+        sort ?? SocketSort.output
+    )
+    const socketTx = await getSocketBridgeTransaction(route)
+    const { allowanceTarget, minimumApprovalAmount } = socketTx.result.approvalData
+    if (socketTx.result.approvalData !== null) {
```
Comment (ken-fun) [inline src/actions/Bridge.ts:43]: If `socketTx.result.approvalData === null` then the previous line would have thrown when trying to destructure `null`. so the destructure should probably happen after the null check.

File: src/actions/FirstClassActions.ts [hunk 10]
```diff
@@ -73,6 +75,17 @@ export abstract class FirstClassActions {
      */
     abstract getAddress(): Promise<Address>
 
+    async bridge(
+        auth: Auth,
+        userId: string,
+        params: BridgeParams,
+        txOptions: EnvOption = (globalThis as any).globalEnvOption
+    ): Promise<Operation> {
+        params.recipient ??= await this.getAddress()
```
Comment (ken-fun) [inline src/actions/FirstClassActions.ts:84]: Better not to mutate objects passed into a function. The caller of this function could be passing in a params object that is re-used later, and would not expect it to change as a result of this call.

File: src/actions/FirstClassActions.ts [hunk 11]
```diff
@@ -73,6 +75,17 @@ export abstract class FirstClassActions {
...
+        const transactionParams = await bridgeTransactionParams(params, await this.getAddress())
+        return await this.createOperation(auth, userId, transactionParams, txOptions)
```
Comment (ken-fun) [inline src/actions/FirstClassActions.ts:86]: There is never a need to `await` in `return await` -- this code is exactly the same as returning the promise directly. The exception is when doing `return await` inside a `try` block.

**Review by ken-fun — APPROVED** [review]

**Review by Chazzzzzzz — APPROVED** [review]

File: src/actions/Bridge.ts (hunk 9 above)
Comment (AlbertSu123) [reply src/actions/Bridge.ts:43]: good catch, fixed!

File: src/actions/FirstClassActions.ts (hunk 10 above)
Comment (AlbertSu123) [reply src/actions/FirstClassActions.ts:84]: Fixed, deep copied the params and modified the recipient on those params

File: src/actions/FirstClassActions.ts (hunk 11 above)
Comment (AlbertSu123) [reply src/actions/FirstClassActions.ts:86]: Fixed! Also, any thoughts on this: https://2ality.com/2015/06/tail-call-optimization.html?

**Pull request merged**